### DISCUSSION TITLE
Test Exercise - React - Mini Social Media Dashboard - 1

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#182] [Tasks]
This app will allow users to create posts, comment on posts, and like posts. 

**Application Overview:**

1. **Posts Feed:** This will display all the posts made by users. Each post will have a "Like" and "Comment" button.

2. **Create Post:** Users will be able to create a new post through a form.

3. **Comments:** Each post will have a section for displaying comments and adding new comments.

Let's break this down step by step.

**Step 1: Setting Up the Project**

Set up a new React project using Create React App. 

```bash
npx create-react-app mini-social-media
cd mini-social-media
npm start
```

**Step 2: Creating the Main `App` Component**

Your `App` component should use state to keep track of the posts in the application. You should initialize it as an empty array. Make sure to import `useState` from React.

**Step 3: Creating the `PostFeed` Component**

Create a `PostFeed` component that accepts `posts` as a prop. This component will map through the posts array, passing each post to a `Post` component (which you'll create next). 

Each post object should have an `id`, `text`, `likes`, and `comments` property. The `comments` property will be an array of comment objects, where each comment object has an `id` and `text`.

**Step 4: Creating the `Post` Component**

Create a `Post` component that accepts a `post` object as a prop. The `Post` component should display the post's text, the number of likes, and have a "Like" button. This component should also display a list of comments using a `Comment` component.

**Step 5: Creating the `Comment` Component**

Create a `Comment` component that accepts a `comment` object as a prop. The `Comment` component should display the comment's text.

**Step 6: Creating the `AddPost` Component**

Create an `AddPost` component. This component should have a form with an input field for the post's text. When the form is submitted, it should call an `onAdd` function (passed as a prop) with an object containing the text of the new post.

**Step 7: Creating the `AddComment` Component**

Similar to `AddPost`, create an `AddComment` component. This component will appear inside each `Post` and allow users to add comments to that specific post.

**Step 8: Implementing the Like Button**

Inside the `Post` component, create a function called `handleLike` which will increment the `likes` property of a post. Make sure that this function updates the state in the `App` component.

**Step 9: Implementing Add Post**

Inside `App`, implement a function called `addPost` that takes a post object and adds it to the `posts` state. Pass this function as a prop to the `AddPost` component.

**Step 10: Implementing Add Comment**

Inside `App`, implement a function called `addComment` that takes a comment object and the `id` of the post. This function should add the comment to the appropriate post in the `posts` state. Pass this function as a prop to the `AddComment` component.

Remember, this application is all about managing state and passing functions as props to manipulate that state. Consider the flow of data through your components, and think about which components should have their own state. Good luck!

**Comment by sam2elect** [comment]
![image](https://github.com/akash-coded/mern/assets/85131631/88449c7e-b337-45c7-a0f0-51a901cab914)

**Comment by chukku10** [comment]
![mini social media](https://github.com/akash-coded/mern/assets/127116126/98d60879-54e2-43bc-b7a7-79af84b0eecc)

**Comment by vk0027** [comment]
![image](https://github.com/akash-coded/mern/assets/123943655/fbdbec57-8f54-423d-b77a-cd8691ee7a88)

**Comment by akash-coded** [comment]
# Solution

Here's how you could build the application based on the steps provided:

```jsx
import React, { useState } from 'react';

function App() {
  const [posts, setPosts] = useState([]);

  const addPost = (text) => {
    const newPost = {
      id: Math.random(), // Generate a random id
      text: text,
      likes: 0,
      comments: [],
    };
    setPosts([newPost, ...posts]);
  };

  const addComment = (text, postId) => {
    setPosts(
      posts.map((post) =>
        post.id === postId
          ? { ...post, comments: [...post.comments, { id: Math.random(), text: text }] }
          : post
      )
    );
  };

  const handleLike = (postId) => {
    setPosts(
      posts.map((post) => (post.id === postId ? { ...post, likes: post.likes + 1 } : post))
    );
  };

  return (
    <div>
      <AddPost onAdd={addPost} />
      <PostFeed posts={posts} onLike={handleLike} onAddComment={addComment} />
    </div>
  );
}

function PostFeed({ posts, onLike, onAddComment }) {
  return (
    <div>
      {posts.map((post) => (
        <Post key={post.id} post={post} onLike={onLike} onAddComment={onAddComment} />
      ))}
    </div>
  );
}

function Post({ post, onLike, onAddComment }) {
  return (
    <div>
      <p>{post.text}</p>
      <p>{post.likes} Likes</p>
      <button onClick={() => onLike(post.id)}>Like</button>
      {post.comments.map((comment) => (
        <Comment key={comment.id} comment={comment} />
      ))}
      <AddComment onAdd={(text) => onAddComment(text, post.id)} />
    </div>
  );
}

function Comment({ comment }) {
  return <p>{comment.text}</p>;
}

function AddPost({ onAdd }) {
  const [text, setText] = useState('');

  const handleSubmit = (e) => {
    e.preventDefault();
    onAdd(text);
    setText('');
  };

  return (
    <form onSubmit={handleSubmit}>
      <input value={text} onChange={(e) => setText(e.target.value)} />
      <button type='submit'>Add Post</button>
    </form>
  );
}

function AddComment({ onAdd }) {
  const [text, setText] = useState('');

  const handleSubmit = (e) => {
    e.preventDefault();
    onAdd(text);
    setText('');
  };

  return (
    <form onSubmit={handleSubmit}>
      <input value={text} onChange={(e) => setText(e.target.value)} />
      <button type='submit'>Add Comment</button>
    </form>
  );
}

export default App;
```
This code covers all the steps mentioned in the task, but keep in mind that in a real-world application, we would separate these components into their own files and organize the project structure in a more scalable way.

This application also lacks a real backend and doesn't handle error states, but it does serve to illustrate how you might create a simple social media-style app with posts, likes, and comments in React.

For simplification, we're generating random IDs for posts and comments using `Math.random()`. In a real-world app, IDs would be generated by your backend/database.

**Comment by sam2elect** [comment]
# UPDATE

```js
import React, { useState } from "react";

function App() {
  const [posts, setPosts] = useState([]);

  const addPost = (text) => {
    const newPost = {
      id: Math.random(), 
      text: text,
      likes: 0,
      comments: [],
    };
    setPosts([newPost, ...posts]);
  };

  const addComment = (text, postId) => {
    setPosts(
      posts.map((post) =>
        post.id === postId
          ? {
              ...post,
              comments: [...post.comments, { id: Math.random(), text: text }],
            }
          : post
      )
    );
  };

  const handleLike = (postId) => {
    setPosts(
      posts.map((post) =>
        post.id === postId ? { ...post, likes: post.likes + 1 } : post
      )
    );
  };

  const handleULike = (postId) => {
    setPosts(
      posts.map((post) =>
        post.id === postId ? { ...post, likes: post.likes - 1 } : post
      )
    );
  };

  const handleDeleteComment = (postId, commentId) => {
    setPosts(
      posts.map((post) =>
        post.id === postId
          ? {
              ...post,
              comments: post.comments.filter(
                (comment) => comment.id !== commentId
              ),
            }
          : post
      )
    );
  };

  return (
    <div>
      <AddPost onAdd={addPost} />
      <PostFeed
        posts={posts}
        onLike={handleLike}
        onUnLike={handleULike}
        onAddComment={addComment}
        onDeleteComment={handleDeleteComment} 
      />
    </div>
  );
}

function PostFeed({ posts, onLike, onUnLike, onAddComment, onDeleteComment }) {
  return (
    <div>
      {posts.map((post) => (
        <Post
          key={post.id}
          post={post}
          onLike={onLike}
          onUnLike={onUnLike}
          onAddComment={onAddComment}
          onDeleteComment={onDeleteComment} 
        />
      ))}
    </div>
  );
}

function Post({ post, onLike, onUnLike, onAddComment, onDeleteComment }) {
  return (
    <div>
      <p>{post.text}</p>
      <p>{post.likes} Likes</p>
      <button onClick={() => onLike(post.id)}>Like</button>
      <button onClick={() => onUnLike(post.id)}>UnLike</button>
      {post.comments.map((comment) => (
        <div key={comment.id}>
          <Comment comment={comment} />
          <button onClick={() => onDeleteComment(post.id, comment.id)}>
            Delete Comment
          </button>
        </div>
      ))}
      <AddComment onAdd={(text) => onAddComment(text, post.id)} />
    </div>
  );
}

function Comment({ comment }) {
  return <p>{comment.text}</p>;
}

function AddPost({ onAdd }) {
  const [text, setText] = useState("");

  const handleSubmit = (e) => {
    e.preventDefault();
    onAdd(text);
    setText("");
  };

  return (
    <form onSubmit={handleSubmit}>
      <input value={text} onChange={(e) => setText(e.target.value)} />
      <button type="submit">Add Post</button>
    </form>
  );
}

function AddComment({ onAdd }) {
  const [text, setText] = useState("");

  const handleSubmit = (e) => {
    e.preventDefault();
    onAdd(text);
    setText("");
  };

  return (
    <form onSubmit={handleSubmit}>
      <input value={text} onChange={(e) => setText(e.target.value)} />
      <button type="submit">Add Comment</button>
    </form>
  );
}

export default App;

```

# ScreenShort
![image](https://github.com/akash-coded/mern/assets/85131631/6de73167-a299-47ed-b38d-31b2db8a61cc)

**Comment by chukku10** [comment]
import React, { useState } from 'react';

function App() {
  const [posts, setPosts] = useState([]);

  const addPost = (text) => {
    const newPost = {
      id: Math.random(), // Generate a random id
      text: text,
      likes: 0,
      comments: [],
    };
    setPosts([newPost, ...posts]);
  };

  const addComment = (text, postId) => {
    setPosts(
      posts.map((post) =>
        post.id === postId
          ? { ...post, comments: [...post.comments, { id: Math.random(), text: text }] }
          : post
      )
    );
  };

  const handleLike = (postId) => {
    setPosts(
      posts.map((post) => (post.id === postId ? { ...post, likes: post.likes + 1 } : post))
    );
  };

  return (
    <div>
      <AddPost onAdd={addPost} />
      <PostFeed posts={posts} onLike={handleLike} onAddComment={addComment} />
    </div>
  );
}

function PostFeed({ posts, onLike, onAddComment }) {
  return (
    <div>
      {posts.map((post) => (
        <Post key={post.id} post={post} onLike={onLike} onAddComment={onAddComment} />
      ))}
    </div>
  );
}

function Post({ post, onLike, onAddComment }) {
  return (
    <div>
      <p>{post.text}</p>
      <p>{post.likes} Likes</p>
      <button onClick={() => onLike(post.id)}>Like</button>
      {post.comments.map((comment) => (
        <Comment key={comment.id} comment={comment} />
      ))}
      <AddComment onAdd={(text) => onAddComment(text, post.id)} />
    </div>
  );
}

function Comment({ comment }) {
  return <p>{comment.text}</p>;
}

function AddPost({ onAdd }) {
  const [text, setText] = useState('');

  const handleSubmit = (e) => {
    e.preventDefault();
    onAdd(text);
    setText('');
  };

  return (
    <form onSubmit={handleSubmit}>
      <input value={text} onChange={(e) => setText(e.target.value)} />
      <button type='submit'>Add Post</button>
    </form>
  );
}

function AddComment({ onAdd }) {
  const [text, setText] = useState('');

  const handleSubmit = (e) => {
    e.preventDefault();
    onAdd(text);
    setText('');
  };

  return (
    <form onSubmit={handleSubmit}>
      <input value={text} onChange={(e) => setText(e.target.value)} />
      <button type='submit'>Add Comment</button>
    </form>
  );
}

export default App;

**Comment by mownica-p** [comment]
![image](https://github.com/akash-coded/mern/assets/126975660/90b4b03e-994b-42d2-897a-71376f84bbf7)

**Comment by sahidraza** [comment]
![Screenshot 2023-07-30 at 10-09-23 React App](https://github.com/akash-coded/mern/assets/114484239/b5604f5b-f107-41dc-9d2c-f8388d04444e)

**Comment by tinapatil13** [comment]
![image](https://github.com/akash-coded/mern/assets/88702583/6ceaf477-ab17-4de6-a034-5ef36191b54c)

**Comment by starvikas1988** [comment]
# Updated with delete comment and unlike implemented.
```js
import React, { useState } from 'react';

function App() {
  const [posts, setPosts] = useState([]);

  const addPost = (text) => {
    const newPost = {
      id: Math.random(), // Generate a random id
      text: text,
      likes: 0,
      comments: [],
    };
    setPosts([newPost, ...posts]);
  };

  const addComment = (text, postId) => {
    setPosts(
      posts.map((post) =>
        post.id === postId
          ? { ...post, comments: [...post.comments, { id: Math.random(), text: text }] }
          : post
      )
    );
  };

  const handleLike = (postId) => {
    setPosts(
      posts.map((post) => (post.id === postId ? { ...post, likes: post.likes + 1 } : post))
    );
  };

  const handleUnlike = (postId) =>{
    setPosts(posts.map((post) => ((post.id === postId && (post.likes >0)) ? { ...post, likes: post.likes - 1 } : post)));
  }

  const handleDeleteComment = (postId, commentId) => {
    setPosts(
      posts.map((post) =>
        post.id === postId
          ? {
              ...post,
              comments: post.comments.filter(
                (comment) => comment.id !== commentId
              ),
            }
          : post
      )
    );
  };

  return (
    <div>
      <AddPost onAdd={addPost} />
      <PostFeed posts={posts} onLike={handleLike} onAddComment={addComment}  
      onDeleteComment={handleDeleteComment}  onUnlike = {handleUnlike}/>
    </div>
  );
}

function PostFeed({ posts, onLike, onAddComment,onDeleteComment ,onUnlike}) {
  return (
    <div>
      {posts.map((post) => (
        <Post key={post.id} post={post} onLike={onLike} onAddComment={onAddComment} 
        onDeleteComment={onDeleteComment} onUnlike = {onUnlike}/>
      ))}
    </div>
  );
}

function Post({ post, onLike, onAddComment,onDeleteComment,onUnlike }) {
  return (
    <div>
      <p>{post.text}</p>
      <p>{post.likes} Likes</p>
      <button onClick={() => onLike(post.id)}>Like</button>
      <button onClick={() => onUnlike(post.id)}>UnLike</button>
      {post.comments.map((comment) => (
       // <Comment key={comment.id} comment={comment} />
      
      <div key={comment.id}>
          <Comment comment={comment} />
          <button onClick={() => onDeleteComment(post.id, comment.id)}>
            Delete Comment
          </button>
        </div>
      ))}
      <AddComment onAdd={(text) => onAddComment(text, post.id)} />
    </div>
  );
}

function Comment({ comment }) {
  return <p>{comment.text}</p>;
}

function AddPost({ onAdd }) {
  const [text, setText] = useState('');

  const handleSubmit = (e) => {
    e.preventDefault();
    onAdd(text);
    setText('');
  };

  return (
    <form onSubmit={handleSubmit}>
      <input value={text} onChange={(e) => setText(e.target.value)} />
      <button type='submit'>Add Post</button>
    </form>
  );
}

function AddComment({ onAdd }) {
  const [text, setText] = useState('');

  const handleSubmit = (e) => {
    e.preventDefault();
    onAdd(text);
    setText('');
  };

  return (
    <form onSubmit={handleSubmit}>
      <input value={text} onChange={(e) => setText(e.target.value)} />
      <button type='submit'>Add Comment</button>
    </form>
  );
}

export default App;
```
# Screenshot
<img width="957" alt="image" src="https://github.com/akash-coded/mern/assets/70792693/a3e478ff-64bb-4512-b08f-18abf353d8b9">

**Comment by Ruchiarth** [comment]
![image](https://github.com/akash-coded/mern/assets/126975575/493ee6bc-65c1-457b-bb46-fc0b836e75ef)
```js
import React, { useState } from "react";

function App() {
  const [posts, setPosts] = useState([]);

  const addPost = (text) => {
    const newPost = {
      id: Math.random(), // Generate a random id
      text: text,
      likes: 0,
      comments: [],
    };
    setPosts([newPost, ...posts]);
  };

  const addComment = (text, postId) => {
    setPosts(
      posts.map((post) =>
        post.id === postId
          ? {
              ...post,
              comments: [...post.comments, { id: Math.random(), text: text }],
            }
          : post
      )
    );
  };

  const handleLike = (postId) => {
    setPosts(
      posts.map((post) =>
        post.id === postId ? { ...post, likes: post.likes + 1 } : post
      )
    );
  };

  return (
    <div>
      <AddPost onAdd={addPost} />
      <PostFeed posts={posts} onLike={handleLike} onAddComment={addComment} />
    </div>
  );
}

function PostFeed({ posts, onLike, onAddComment }) {
  return (
    <div>
      {posts.map((post) => (
        <Post
          key={post.id}
          post={post}
          onLike={onLike}
          onAddComment={onAddComment}
        />
      ))}
    </div>
  );
}

function Post({ post, onLike, onAddComment }) {
  return (
    <div>
      <p>{post.text}</p>
      <p>{post.likes} Likes</p>
      <button onClick={() => onLike(post.id)}>Like</button>
      {post.comments.map((comment) => (
        <Comment key={comment.id} comment={comment} />
      ))}
      <AddComment onAdd={(text) => onAddComment(text, post.id)} />
    </div>
  );
}

function Comment({ comment }) {
  return <p>{comment.text}</p>;
}

function AddPost({ onAdd }) {
  const [text, setText] = useState("");

  const handleSubmit = (e) => {
    e.preventDefault();
    onAdd(text);
    setText("");
  };

  return (
    <form onSubmit={handleSubmit}>
      <input value={text} onChange={(e) => setText(e.target.value)} />
      <button type="submit">Add Post</button>
    </form>
  );
}

function AddComment({ onAdd }) {
  const [text, setText] = useState("");

  const handleSubmit = (e) => {
    e.preventDefault();
    onAdd(text);
    setText("");
  };

  return (
    <form onSubmit={handleSubmit}>
      <input value={text} onChange={(e) => setText(e.target.value)} />
      <button type="submit">Add Comment</button>
    </form>
  );
}

export default App;
```
